### ISSUE TITLE
"warning_capacity": "No channels have sufficient incoming capacity"

### ISSUE THREAD

**Issue by Coinomatron** [ElementsProject/lightning#2314]
Error when making an invoice. I'm probably not technical enough to see the solution, I have no idea how to fix this. I have 1 channel open, from which I've made payments.

**Comment by rustyrussell** [comment]
It means there isn't enough incoming capacity in your 1 channel for someone to pay the invoice.  listchannels will show the channel balance.

**Comment by Coinomatron** [comment]
It's an invoice for 1400 sats, the channel has 59712 left out of 60000 sats. Does it mean that the difference has to be equal or bigger than 1400 sats ?

**Comment by conscott** [comment]
@Coinomatron  - yes if you run `lightning-cli listpeers`, your incoming channel balance is:
`msatoshi_total` - `msatoshi_to_us`, while your outgoing balance is simply `msatoshi_to_us`

Your can increase your incoming capacity by having someone open a channel with your node, or by simply spending more money on your existing channel

**Comment by Coinomatron** [comment]
Is there a way to move funds from the channel that is too 'full' to another channel I have opened myself? I couldn't find something like that in the help commands.

**Comment by cdecker** [comment]
> Is there a way to move funds from the channel that is too 'full' to another channel I have opened myself? I couldn't find something like that in the help commands.

For that you need to create a circular route. There is currently no tool for that. But as long as you have one channel with sufficient capacity you should not get that warning anymore.

**Comment by Coinomatron** [comment]
Ok, so basically I need to spend an amount to receive that equal amount if I have 1 channel that's too 'full'?

**Comment by cdecker** [comment]
Yep, that's basically it, you send out one way, and receive on the other.

**Comment by Coinomatron** [comment]
Ok, closing, but would be nice to have some other solution to this than spending funds not to ones own channel(s).

**Comment by Coinomatron** [comment]
One more thing, if people want to use lightning only to receive payments (for instance twitter bot), how would this issue be resolved?

**Comment by conscott** [comment]
Anyone can still send you payments by opening (and funding) a channel with your node, therefore creating incoming capacity the rest of the network can route through. Your site can give the nodeid/ip/port information necessary to do this, or you can simply ask other well connected nodes (like people on irc/reddit/twitter) to open a channel with you.

**Comment by Coinomatron** [comment]
Ah, ok. Thanks!

**Comment by billygarrison** [comment]
> For that you need to create a circular route. There is currently no tool for that.

@cdecker is it currently possible to build a tool for this using pylightning / lightning-rpc? I'd be happy to attempt it but would like to know if there are any pitfalls I should be aware of.

**Comment by cdecker** [comment]
> @cdecker is it currently possible to build a tool for this using pylightning / lightning-rpc? I'd be happy to attempt it but would like to know if there are any pitfalls I should be aware of.

Oh, absolutely you can export the entire network topology using `listchannels` and then try to find cycles around you. I can definitely help if you need help, and if not, let me know how it turns out, I'd be interested as well :-)

**Comment by conscott** [comment]
fyi @billygarrison  - there is something similiar for `lnd` here: https://github.com/C-Otto/rebalance-lnd

it's not super portable to c-lightning, but if you understand the logic, a similar tool could be made.

**Comment by billygarrison** [comment]
Thanks to you both. I'll be sure to reach out if I need help.

**Comment by Coinomatron** [comment]
I've opened a channel from asinq on android to my c-lightning (desktop) node, hoping it would get more incoming capacity. Instead, I'm seeing that the channel funded by asinq has the same limitation on the c-lightning side. (so it still can't receive more than the asinq side has spent, since the c-lightning side is showing the same amounts. Below c-lightning output of "listpeers".

"msatoshi_to_us": 4254589, 
"msatoshi_to_us_min": 0, 
"msatoshi_to_us_max": 4254589, 
"msatoshi_total": 191058000,

**Comment by SimonVrouwe** [comment]
I think here lies the problem

https://github.com/ElementsProject/lightning/blob/90f9985dd209a6f2b0fc89b7c5c1c8e341229030/lightningd/invoice.c#L200-L204

If we have only a few (or just one) incoming channel, the probability of selecting (that) one seems small. And then it complains not enough incoming capacity. Wiki page link in the code about reservoir sampling states that it is used _very large_   **n**, but the number of incoming channels is in most cases just a few?

Maybe just change above into this?:
```
/* Take the channel with largest excess (incoming) capacity */
               if (excess.millisatoshis > wsum_max) {
                       wsum_max = excess.millisatoshis;
                        r = &inchans[i];
               }
```
Simply pick the one with the highest incoming capacity, assumed we only supply one route-hint.

**Comment by rustyrussell** [comment]
I don't think so.  For the first channel, `wsum = excess + 1`.  Thus, it's always selected, unless we get overflow (which is a real bug, but not the issue here AFAICT).

The problem is two-fold:
1. Eclair opens a private channel, and we have a public channel, so we don't expose it by default.
2. Even if we did `expostprivatechannels=true` we *still* won't use that channel, because we know
    know that node doesn't have any public channels, so telling people about it is pretty useless.

**Comment by sangbida** [comment]
This issue is being closed as part of issue cleanup.

**Reason**: Very old issue (7.0 years old) no recent activity

If you believe this issue should remain open, please comment and we can reopen it.
